src/_guides/language/effective-dart/design.md

```diff
@@ -1005,90 +1089,254 @@ var names = people.map((Person person) {
 });
 {% endprettify %}
 
+In rare cases, the surrounding context is not precise enough to provide a type
+for one or more of the function's parameters. In those cases, you may need to
+annotate.
+
 
-### AVOID annotating with `dynamic` when not required.
+### PREFER annotating with `dynamic` even when not required.
 
-{% comment %}
-update-for-dart-2.0
-https://github.com/dart-lang/site-www/pull/480#discussion_r157355593
-{% endcomment %}
+When you omit a type annotation, *most* of the time, Dart infers a correct
+static type for you. That word "most" is the alarming part. Sometimes, inference
+fails and silently uses `dynamic` instead. If you didn't realize that inference
+failed, you now have code that *looks* statically safe but actually has no
+type checking whatsoever.
 
-In most places in Dart, a type annotation can be omitted, in which case the type
-will automatically be `dynamic`. Thus, omitting the type annotation entirely is
-semantically equivalent but more terse.
+Sometimes, `dynamic` is what you want. Dart does support a dynamic type for good
+reasons. But a casual reader of your code who sees an annotation is missing has
+no way of knowing if you intended it to be `dynamic`, expected inference to fill
+in some other type, or simply forgot to write the annotation.
+
+When `dynamic` is the type you want, using it explicitly makes your intent
+clear.
 
 {:.good-style}
-<?code-excerpt "misc/lib/effective_dart/design_good.dart (avoid-dynamic)"?>
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (prefer-dynamic)"?>
 {% prettify dart %}
-lookUpOrDefault(String name, Map map, defaultValue) {
-  var value = map[name];
-  if (value != null) return value;
-  return defaultValue;
-}
+dynamic mergeJson(dynamic original, dynamic changes) => ...
+{% endprettify %}
+
+{:.bad-style}
+<?code-excerpt "misc/lib/effective_dart/design_bad.dart (prefer-dynamic)"?>
+{% prettify dart %}
+mergeJson(original, changes) => ...
+{% endprettify %}
+
+
+<aside class="alert alert-info" markdown="1">
+
+Before Dart 2, this guideline stated the exact opposite: *don't* annotate with
```

Comment:
I earlier suggested having a **Changed guideline:** intro to notes like these. Do you not want to call out changes to guidelines? **New guideline** might also be a handy thing for people to search for. Or maybe you could just post a news/medium article about what's changed, and we can refer people to that.

Reply:
I don't generally want to call them out, no. Most readers of the guidelines probably don't care about its tumultuous history. I only called out this one because it is an exact inversion of what it previously stated. In most other cases, the guidelines are simply new or only slightly tweaked in intent.